linkml/generators/fhirgen.py

```diff
@@ -0,0 +1,393 @@
+            for a in attrs:
+                c.attributes[a.name] = a
+            c.slots = []
+            return yaml_dumper.dumps(c)
+
+    def all_enum_objects(self) -> Iterator[EnumDefinition]:
+        """
+        all enum objects in schema
+
+        Ensures rank is non-null
+        :return: iterator
+        """
+        elts = self.schemaview.all_enums().values()
+        _ensure_ranked(elts)
+        for e in elts:
+            yield e
+
+    def class_hierarchy_as_tuples(self) -> Iterator[Tuple[int, ClassDefinitionName]]:
+        """
+        Generate a hierarchical representation of all classes in the schema
+
+        This is represented as a list of tuples (depth: int, cls: ClassDefinitionName),
+        where the order is pre-order depth first traversal
+
+        This can then be used to draw a hierarchy within jinja in a number of ways; e.g
+
+        - markdown (by using indentation of " " * depth on a list)
+        - tables (by placing fake indentation e.g using underscores)
+
+        Simply iterate through all tuples, drawing each in the appropriate way
+
+        Note: By default all classes are ordered alphabetically
+
+        :return: tuples (depth: int, cls: ClassDefinitionName)
+        """
+        sv = self.schemaview
+        roots = sv.class_roots(mixins=False)
+
+        # by default the classes are sorted alphabetically
+        roots = sorted(roots, key=str.casefold, reverse=True)
+
+        # the stack holds tuples of depth-class that have still to be processed.
+        # we seed this with all root classes (which have depth 0)
+        # note the stack is processed from the last element first, ie. LIFO
+        stack = [(0, root) for root in roots]
+        # use iterative depth first traversal
+        while len(stack) > 0:
+            depth, class_name = stack.pop()
+            yield depth, class_name
+            children = sorted(
+                sv.class_children(class_name=class_name, mixins=False),
+                key=str.casefold,
+                reverse=True,
+            )
+            for child in children:
+                # depth first - place at end of stack (to be processed next)
+                stack.append((depth + 1, child))
+
+
+# todo: Create an output file or just print to stdout as it is now?
+# TODO: test is generating an error: generator.decorator: default=g.valid_formats[0],
+# @shared_arguments(FhirValueSetGenerator)
+@click.option(
+    "--outpath",
+    "-o",
+    required=True,
+    help="Path for output file",
+)
+@click.command()
+# def cli(yamlfile, outpath, **args):
+def cli(yamlfile, outpath, **args):
+    """Generate FHIR ValueSet JSON from a LinkML YAML Enum schema"""
+    # gen = FhirValueSetGenerator(
+    FhirValueSetGenerator(
+        yamlfile,
+        outpath=outpath,
+        **args,
+    )
+    # todo: print() was from docgen. Do I want to print to stdout? What if multiple files?
+    # print(gen.serialize())
```

Comment:
@hrshdhgd I saw `print(gen.serialize())` in `docgen`. Is there a reason for it? I was thinking it could be useful to print the JSON file to `stdout`, but I think this is not a good idea for reasons: (i) multiple files being generated at once, (ii) only useful when using LinkML in the CLI to do something like `linkml-fhirgen myLinkml.yml > myFhir.json`, but not useful when using the Python API.

Reply:
OK, this is one of the confusing things about the old-school visitor-pattern based generators

check this:

https://github.com/linkml/linkml/blob/596bedfb0b5a3abd611fa627aaff04c3ae19510b/linkml/utils/generator.py#L144-L145

Yes, that is kind of horrible

for the new style schemaview-based generators there should be no need for this malarkey

Reply:
Ahh, I see. After stripping away `docgen` of what I don't need, it doesn't look like schemaview is being used by the serialization or write functions. I probably will strip references to it away entirely. But if there's something I about it that I ought to use, let me know.